routing/rt/graph/graph.go

```diff
@@ -0,0 +1,536 @@
+// Copyright (c) 2016 Bitfury Group Limited
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php
+
+package graph
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"math/rand"
+	"os"
+	"strconv"
+	"bytes"
+)
+
+var (
+	CyclicDataError         = errors.New("Cyclic data")
+	PathNotFoundError       = errors.New("Path not found")
+	NodeNotFoundError       = errors.New("Node not found")
+	EdgeNotFoundError       = errors.New("Edge not found")
+)
+
+type ID struct {
+	ID [33]byte
+}
+
+var NilID = ID{ID: [33]byte{}}
+
+func NewIDDebug(s interface{}) ID {
+	switch id := s.(type) {
+	case string:
+		x := ID{}
+		copy(x.ID[:], []byte(id))
+		return x
+	case []byte:
+		x := ID{}
+		copy(x.ID[:], id)
+		return x
+	case [32]byte:
+		return NewIDDebug(id[:])
+	case int:
+		return NewIDDebug(strconv.Itoa(id))
+	default:
+		return NewIDDebug(fmt.Sprintf("%v", id))
+	}
+}
+
+func NewID(s [33]byte) ID{
+	return ID{ID:s}
+}
+
+func (s ID) String() string {
+	return string(s.ID[:])
+}
+
+func (s ID) ToByte32() [32]byte {
+	var rez [32]byte
+	copy(rez[:], s.ID[:])
+	return rez
+}
+
+func (s ID) ToByte33() [33]byte {
+	var rez [33]byte
+	copy(rez[:], s.ID[:])
+	return rez
+}
+
+func (s ID) ToByte() []byte {
+	return s.ID[:]
+}
+
+func (s ID) IsEmpty() bool {
+	var z [33]byte
+	return bytes.Equal(s.ID[:], z[:])
+}
+
+func (s ID) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + s.String() + `"`), nil
+}
+
+type EdgeID struct {
```

Comment:
Edge already contains additional information in ChannelInfo. You suggest merging EdgeID and ChannelInfo into one structure Edge. It seems good for me to distinguish edge identifier(EdgeID) from channel(edge) information (ChannelInfo). It is better to have distinct type EdgeID for edge identification instead of assuming that it is wire.OutPoint. In future, we may want to switch from hash to <block height>+<transaction number> as was discussed in Milan.